### ISSUE TITLE
Spawner instance reused after a failed start

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/jupyterhub#3592]
If I'm not mistaken, each server starting should get a fresh Spawner instance for each spawn attempt.

I made a spawn attempt that failed because start_timeout triggered after in this case, KubeSpawner, had failed to schedule a k8s pod on a k8s node. After this attempt, I tried to start another server with a different configuration via KubeSpawner's profile_list, but I ended up with a mixed configuration where one setting came from the first spawn attempt and the other came from the new spawn attempt. This to me indicated that the Spawner instance had been re-used.

**Comment by minrk** [comment]
Yes, indeed! I'm not sure how this can happen, but more details would help. Maybe logs. Each spawn should indeed create a new Spawner.

Is it possible the spawner was stuck in 'pending' state instead of actually finished failing?

**Comment by consideRatio** [comment]
I'm not sure about the spawner instance's state, but the k8s pod was stuck in a Pending state but must have been terminated between the attempts as otherwise the new pod couldn't startup I think.

I have no logs to provide at this point. I think the reproduction procedure would be like this.

Reproduction setup:
1. Configure a short start_timeout to avoid needing to wait for a long time.
2. Configure a pre_spawn_hook that updates the node_selector to include a node label that doesn't exist
3. Configure the pre_spawn_hook to make an assertion that the node_selector shouldn't be set before it sets it. If it is, that means that this logic has already run for this spawner instance.

Reproduction steps:
1. Login to jupyterhub
2. Spawn a pod. This should fail with a timeout as the pod should be stuck in a pending state
3. Spawn a pod again. This should fail with the broken assertion if it reproduces.

**Comment by consideRatio** [comment]
This seems like a potential cause of this issue @minrk. Any ideas of a suggested change comes to mind?

https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L945-L954

**Comment by minrk** [comment]
@consideRatio good eye! I think you're right that this is going to be the root cause. In most cases, failure to dispose of a previous spawner isn't an issue, but certain cases, especially with dynamic options, certainly will not fully reinitialize unless a new Spawner is instantiated. Should search Issues, I think this may be related to some other reports around spawners having 'stale' state.

Will need to think carefully about exactly when to dispose of a failed Spawner vs create a fresh one. We probably want these same checks for `spawner._spawn_future` in User.spawn, but possibly elsewhere as well (when showing spawn form, for example).
